core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c

```diff
@@ -321,6 +325,9 @@ struct fd_object {
     struct refcount refcount;
     __wasi_filetype_t type;
     int number;
+#ifdef WAMR_SGX_IPFS
+    void *sgx_file;
+#endif
```

Comment:
@JamesMenetrey It may be not a very good option to modify this file by adding macro controls, in our design, the related system APIs should be provided by the platform layer. As you have implemented the `ipfs_xxx` APIs in sgx_ipfs.c/h, shall we encapsulate a layer to re-implement the sgx platform APIs e.g. fopen/fclose/seek/.., for example, create a map and the operations between file descriptor (fd) and the file handle of sgx ipfs:
```
typedef struct {
    void *ipfs_file;
    uint32 ipfs_fd;
    ...
} IPFSFileMap;

IPFSFileMap *ipfs_file_list;            //Or an array with max size, or a hash map
os_corp_mutex ipfs_file_list_lock; //lock for multi-thread

int ipfs_init()  // To initialize the list, and call this in bh_platform_init()
{...
}
void ipfs_destroy() // To destroy the list, and call this in bh_platform_destroy()
{...
}

void * ipfs_fd2file(int ipfs_fd)
{...
}

int ipfs_file2fd(void *ipfs_file)
{...
}
```

And provide operations to allocate file descript for new file handle,
for API read in sgx_file.c, we may rewrite it like:
```C
ssize_t
read(int fd, void *buf, size_t size)
{
    ssize_t ret;
    int size_read_max = 2048, size_read, total_size_read = 0, count, i;
    char *p = buf;

#if WASM_ENABLE_SGX_IPFS != 0
   ipfs_file = ipfs_fd2file(fd);
   if (!ipfs_file) { report error }
   operate on the file handle
#endif

    if (buf == NULL) {
        TRACE_FUNC();
        return -1;
    }
    ...
}
```

Reply:
Hey, thanks for the insights! I have refactored to implement all the IPFS hooks in the SGX dedicated platform file. I opted to use a hashmap to map the file descriptors to the SGX IPFS pointers, so the complexity of pointers resolution is O(1). I did not use a mutex manually in my code, since the hashmap provides this ability by setting the right argument in the constructor.